h2o-core/src/main/java/hex/grid/HyperSpaceWalker.java

```diff
@@ -437,6 +440,123 @@ public void modelFailed(Model failedModel) {
+                                    ArrayList<Function<HashMap<String, Object>, HashMap<String, Object>>> filterFunctions) {
+      super(params, hyperParams, paramsBuilderFactory, search_criteria);
+    }
+
+    @Override
+    public HyperSpaceIterator<MP> iterator() {
+      ArrayList<HashMap<String, Object>> allElements = new ArrayList<>();
+
+      _hyperParams.forEach((hpKey, hpValues) -> {
+        if(allElements.isEmpty()) {
+          Arrays.stream(hpValues).forEach(hpValue -> {
+            HashMap<String, Object> newGridItem = new HashMap<>();
+            newGridItem.put(hpKey, hpValue);
+            allElements.add(newGridItem);
+          });
+        } else {
+          Stream<HashMap<String, Object>> expandedResult = allElements.stream().flatMap(existingItem -> {
+            return Arrays.stream(hpValues).map(hpValue -> {
+              HashMap<String, Object> clone = (HashMap<String, Object>) existingItem.clone();
+              clone.put(hpKey, hpValue);
+              return clone;
+            });
+          });
+          ArrayList<HashMap<String, Object>> collect = expandedResult.collect(Collectors.toCollection(ArrayList::new));
+          allElements.clear();
+          allElements.addAll(collect);
+        }
+      });
+      Collections.shuffle(allElements);
+
+      // TODO Apply `filterFunctions` here to `allElements`
+              
+      Iterator<Object[]> iterator = allElements.stream().map(item -> Arrays.stream(_hyperParamNames).map(item::get).toArray()).collect(Collectors.toCollection(ArrayList::new)).iterator();
+
+      return new HyperSpaceIterator<MP>() {
+        /** Hyper params permutation.
+         */
+
+        @Override
+        public MP nextModelParameters(Model previousModel) {
+          if (hasNext(previousModel)) { // or just iterator.hasNext()
+            // Get clone of parameters
+            MP commonModelParams = (MP) _params.clone();
+            // Fill model parameters
+            MP params = getModelParams(commonModelParams, getCurrentRawParameters()); // or just iterator.next()
+
+            return params;
+          } else {
+            throw new NoSuchElementException("No more elements to explore in hyper-space!");
+          }
+        }
+
+        @Override
+        public boolean hasNext(Model previousModel) {
+          return iterator.hasNext();
+        }
+
+        @Override public void reset() {
+          //TODO
+        }
+
+        @Override
+        public double time_remaining_secs() { return Double.MAX_VALUE; }
+
+        @Override
+        public double max_runtime_secs() { return Double.MAX_VALUE; }
+
+        public int max_models() { return _maxHyperSpaceSize > Integer.MAX_VALUE ? Integer.MAX_VALUE : (int)_maxHyperSpaceSize; }
+
+        @Override
+        public void modelFailed(Model failedModel) {
+          // nada
+        }
+
+        @Override
+        public Object[] getCurrentRawParameters() {
```

Comment:
just to make sure - is this correct? this method gets called when the model building fails

Reply:
@michalkurka most likely it is not correct, and taking into account name of the method we should not return next element but last one, i.e. current one. Thanks!